### PR TITLE
WFLY-12681 Provide a 'microprofile-platform' Galleon layer

### PR DESCRIPTION
**Pull request by rhusar**
Resolves
https://issues.redhat.com/browse/WFLY-12681

Proposal
https://github.com/wildfly/wildfly-proposals/pull/293

### CODE REVIEW

File: docs/src/main/asciidoc/_admin-guide/Galleon_provisioning.adoc [hunk 1]
```diff
@@ -122,6 +122,7 @@ Basic layers:
 * _jaxrs_: Support for JAXRS. Depends on _web-server_ (defined in the WildFly servlet feature-pack).
 * _jms-activemq_: Support for connections to a remote JMS broker.
 * _jpa_: Support for JPA (using the latest WildFly supported Hibernate release).
+* _microprofile_: Support for MicroProfile specification including Config, Fault Tolerance, JWT, Metrics, OpenAPI, OpenTracing.
```
Comment (pferraro) [inline docs/src/main/asciidoc/_admin-guide/Galleon_provisioning.adoc]: I'm not sure I would bother enumerating the individual specs, unless we want to keep this list updated (e.g. health) with every new release.

**Review by pferraro — CHANGES_REQUESTED** [review]

File: docs/src/main/asciidoc/_admin-guide/Galleon_provisioning.adoc (hunk 1 above)
Comment (rhusar) [reply docs/src/main/asciidoc/_admin-guide/Galleon_provisioning.adoc]: My thinking was to keep the list updated so people dont assume this includes some specs we dont yet support or ones that might be only incubating. What better wording do you suggest?

Comment (darranl) [reply docs/src/main/asciidoc/_admin-guide/Galleon_provisioning.adoc]: Would including an additional qualifier in the name make sense?  Either platform version  or allow for a second one `microprofile_experimental`

Comment (rhusar) [reply docs/src/main/asciidoc/_admin-guide/Galleon_provisioning.adoc]: Reworded to be more general.

**Comment by rhusar** [comment]
As per @bstansberry 's suggestion, made all the specs optional except for config which is required by all specifications.

**Comment by rhusar** [comment]
For the sake of completeness, I have opened the proposal as well https://github.com/wildfly/wildfly-proposals/pull/293

**Review by pferraro — APPROVED** [review]
LGTM

**Comment by rhusar** [comment]
Rebased again..

**Review by darranl — APPROVED** [review]

**Comment by darranl** [comment]
No objection from me, do we need to double check the approach on Monday?

Also FYI this needs a rebase again.

**Comment by rhusar** [comment]
@darranl Right, lets discuss on Mondays call.

File: galleon-pack/src/main/resources/layers/standalone/microprofile/layer-spec.xml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  ~ 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+  -->
+<layer-spec xmlns="urn:jboss:galleon:layer-spec:1.0" name="microprofile">
+    <dependencies>
+        <layer name="microprofile-config" optional="true"/>
+        <layer name="microprofile-fault-tolerance" optional="true"/>
+        <layer name="microprofile-jwt" optional="true"/>
+        <layer name="microprofile-metrics" optional="true"/>
+        <layer name="microprofile-openapi" optional="true"/>
+        <layer name="open-tracing" optional="true"/>
```
Comment (yersan) [inline galleon-pack/src/main/resources/layers/standalone/microprofile/layer-spec.xml:30]: @rhusar Just out of curiosity, why we don't include microprofile-health?

Comment (rhusar) [reply galleon-pack/src/main/resources/layers/standalone/microprofile/layer-spec.xml:31]: @yersan Fixed, thanks. The reason being that 'microprofile-health' layer did not exist in January.

**Comment by darranl** [comment]
Sorry @rhusar but this is conflicting again.

**Comment by rhusar** [comment]
@darranl Thanks, back from PTO, rebased now.

File: testsuite/layers/pom.xml [hunk 3]
```diff
@@ -2266,6 +2302,7 @@
                                                 <layer>logging</layer>
                                                 <layer>management</layer>
                                                 <layer>management-security-realm</layer>
+                                                <layer>microprofile-platform</layer>
                                                 <layer>microprofile-config</layer>
```
Comment (tommaso-borgato) [inline testsuite/layers/pom.xml:2306]: @rhusar given `microprofile-platform` already includes `microprofile-config`, `microprofile-fault-tolerance`, `microprofile-health`, `microprofile-jwt`, `microprofile-metrics`, `microprofile-openapi`, `open-tracing`: can we remove them from the list? or we keep them to prevent errors in case, in some future, they are removed from `microprofile-platform` or `microprofile-platform` itself is removed?

**Review by tommaso-borgato — CHANGES_REQUESTED** [review]

File: testsuite/layers/pom.xml (hunk 3 above)
Comment (rhusar) [reply testsuite/layers/pom.xml:2306]: @tommaso-borgato We can – but the existing layers are already overlapping (e.g. undertow is listed explicitly even though its a dependency of any of the web server layers, e.g. datasources-web-server). So this is intentional.

File: testsuite/integration/basic/pom.xml [hunk 4]
```diff
@@ -1229,6 +1270,45 @@
...
+                                        <JBOSS_HOME>${project.build.directory}/wildfly-microprofile-platform</JBOSS_HOME>
+                                        <!-- dummy env property for MicroProfile Config testing -->
+                                        <MPCONFIG_TEST_ENV_VAR>integ-basicmpconfig-test-env-var-value</MPCONFIG_TEST_ENV_VAR>
+                                        <TEST_PROP>JMXPropertyEditorsTestCase-WFLY-12899</TEST_PROP>
+                                    </environmentVariables>
+                                    <redirectTestOutputToFile>${testLogToFile}</redirectTestOutputToFile>
+                                    <includes>
+                                        <include>org/wildfly/test/integration/microprofile/**/*TestCase.java</include>
+                                    </includes>
+                                    <excludes>
```
Comment (tommaso-borgato) [inline testsuite/integration/basic/pom.xml:1299]: @rhusar what drove the decision to exclude these tests? I am asking just for getting some understanding ... (I already asked on the AD, so if you answer there I'll copy the answer here and vice-versa)

**Review by tommaso-borgato — CHANGES_REQUESTED** [review]

File: testsuite/integration/basic/pom.xml (hunk 4 above)
Comment (rhusar) [reply testsuite/integration/basic/pom.xml:1299]: @tommaso-borgato These tests have other dependencies in order to run (e.g. security) while the profile only provisions base server (`jaxrs-server`) with the `microprofile-plaform` layer and of course `jmx-remoting` & `security-manager` which are needed by Arquillian. We might consider adding more layers to support more tests, but I am not sure if that doesn't defeat the purpose of the test.

Comment (tommaso-borgato) [reply testsuite/integration/basic/pom.xml:1299]: @rhusar no, that's fine .... it was just for understanding

File: galleon-pack/src/main/resources/layers/standalone/microprofile-platform/layer-spec.xml [hunk 5]
```diff
@@ -0,0 +1,33 @@
+<?xml version="1.0" ?>
```
Comment (tommaso-borgato) [inline galleon-pack/src/main/resources/layers/standalone/microprofile-platform/layer-spec.xml:1]: @rhusar please consider reading [this comment](https://docs.google.com/document/d/1FXdEQc1Q8v1GmsTbwb6XDl_GTFrcFYuQPpkAOV6hrUI/edit?disco=AAAAJxVMytU) in order to validate the location where this layer is defined

File: testsuite/integration/basic/pom.xml (hunk 4 above)
Comment (tommaso-borgato) [reply testsuite/integration/basic/pom.xml:1299]: @rhusar .... this layer looks like one of the biggest candidate for adoption, also given we are in the XP 2 context; 

what about adding one or more galleon-maven-plugin & maven-surefire-plugin executions in order to run the excluded tests on wildfly servers taylored with the minimal number of necessary layers? Something analogous will be done for the JSF layer with regard to the EJB layer (@yersan is that correct?) ...

**Review by tommaso-borgato — CHANGES_REQUESTED** [review]

File: galleon-pack/src/main/resources/layers/standalone/microprofile-platform/layer-spec.xml (hunk 5 above)
Comment (rhusar) [reply galleon-pack/src/main/resources/layers/standalone/microprofile-platform/layer-spec.xml:1]: @tommaso-borgato The comment confirms the location is correct. Resolving.

File: testsuite/integration/basic/pom.xml (hunk 4 above)
Comment (rhusar) [reply testsuite/integration/basic/pom.xml:1299]: @tommaso-borgato We can certainly do that, the price we pay is the extra testsuite execution time which is kind of exploding atm. Do you have a link handy where the JSF+EJB layer tests are discussed by any chance?

Comment (tommaso-borgato) [reply testsuite/integration/basic/pom.xml:1299]: @rhusar I am waiting for EAP7-1529 to be completed with JSF+EJB test, that's why I tagged @yersan: here is the TP snippet relative to this: https://gitlab.mw.lab.eng.bos.redhat.com/jbossqe-eap/test-plans/merge_requests/380/diffs#4e1bf40d542063d096fcd61d1fb6f49b4aba74ab_0_177

Probably he can contribute to this discussion ....

Comment (tommaso-borgato) [reply testsuite/integration/basic/pom.xml:1299]: @rhusar in addition to that, microprofile tests are fast (about 10 seconds on my laptop)

Comment (tommaso-borgato) [reply testsuite/integration/basic/pom.xml:1299]: ^^^ do you have any update on the missing tests? would it be possible, for example, to add security tests on a slimmed server with the minimum amount of necessary layers?

**Review by tommaso-borgato — CHANGES_REQUESTED** [review]

File: testsuite/integration/basic/pom.xml (hunk 4 above)
Comment (rhusar) [reply testsuite/integration/basic/pom.xml:1295]: Actually, it looks like the problem is only with the test setup. I will put this on the backlog, but its not clear what those tests are missing – will need to check with Darran when he comes back.

**Comment by rhusar** [comment]
@bstansberry Needed a rebase again (also I think I botched the previous rebased, fixed that too).
